### PR TITLE
[wip] added option to handle OAuth grants at a per-client granularity

### PR DESCRIPTION
**Pull request by stevekuznetsov**
@liggitt PTAL

How do we test this?

### CODE REVIEW

File: pkg/auth/api/types.go [hunk 1]
```diff
@@ -40,6 +40,7 @@ type Client interface {
 	GetSecret() string
 	GetRedirectUri() string
 	GetUserData() interface{}
```
Comment (liggitt) [inline pkg/auth/api/types.go]: pretty sure this was made to match the osin client interface... we'll likely need to include the grant strategy in the "user data"... that's our bucket to plumb whatever data we want through all the osin apis

File: pkg/cmd/server/api/v1/types.go [hunk 2]
```diff
@@ -787,6 +787,8 @@ const (
 	GrantHandlerPrompt GrantHandlerType = "prompt"
 	// GrantHandlerDeny auto-denies client authorization grant requests
 	GrantHandlerDeny GrantHandlerType = "deny"
+	// GrantHandlerPerClient determines behavior on a per-client basis
+	GrantHandlerPerClient GrantHandlerType = "perclient"
```
Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: can drop this

File: pkg/oauth/server/osinserver/registrystorage/storage.go [hunk 3]
```diff
@@ -63,6 +63,10 @@ func (w *clientWrapper) GetUserData() interface{} {
 	return w.client
```
Comment (liggitt) [inline pkg/oauth/server/osinserver/registrystorage/storage.go]: we're already returning the client object here... we should be able to get the grant strategy from it

Comment (stevekuznetsov) [inline pkg/oauth/server/osinserver/registrystorage/storage.go]: Ah, ok. will do

File: pkg/auth/oauth/handlers/grant.go [hunk 4]
```diff
@@ -122,3 +125,41 @@ func (g *redirectGrant) GrantNeeded(user user.Info, grant *api.Grant, w http.Res
 	http.Redirect(w, req, redirectURL.String(), http.StatusFound)
 	return false, true, nil
 }
+
+// NewPerClientGrant returns a grant handler that determines what to do based on the grant strategy in the client
+func NewPerClientGrant(redirect GrantHandler, defaultStrategy string) GrantHandler {
+	return &perClientGrant{
+		auto:            NewAutoGrant(),
+		redirect:        redirect,
```
Comment (liggitt) [inline pkg/auth/oauth/handlers/grant.go]: call this prompt, not redirect

**Comment by stevekuznetsov** [comment]
@liggitt addressed comments

**Comment by stevekuznetsov** [comment]
doc at https://github.com/openshift/openshift-docs/pull/1682

**Comment by stevekuznetsov** [comment]
@liggitt @smarterclayton how has running `hack/update-generated-conversion.sh` left me in some sort of indeterminate state where re-running the script fails?
```
$ hack/update-generated-conversions.sh 
Generating for version v1beta3
# github.com/openshift/origin/pkg/api/v1
pkg/api/v1/conversion_generated.go:4805: undefined: serverapiv1 in serverapiv1.GrantHandlerType
pkg/api/v1/conversion_generated.go:5008: undefined: serverapi in serverapi.GrantHandlerType
```

**Comment by stevekuznetsov** [comment]
Do we not support one `api` package importing another? Will we need to duplicate the `GrantHandlerType` code?

File: pkg/api/v1beta3/conversion_generated.go [hunk 5]
```diff
@@ -6684,8 +6588,6 @@ func init() {
 		autoConvert_api_OAuthAuthorizeToken_To_v1beta3_OAuthAuthorizeToken,
 		autoConvert_api_OAuthClientAuthorizationList_To_v1beta3_OAuthClientAuthorizationList,
 		autoConvert_api_OAuthClientAuthorization_To_v1beta3_OAuthClientAuthorization,
-		autoConvert_api_OAuthClientList_To_v1beta3_OAuthClientList,
```
Comment (stevekuznetsov) [inline pkg/api/v1beta3/conversion_generated.go]: @liggitt do we want custom conversion funcs for these?

File: pkg/oauth/api/v1/types.go [hunk 6]
```diff
@@ -79,8 +79,23 @@ type OAuthClient struct {
 
 	// RedirectURIs is the valid redirection URIs associated with a client
 	RedirectURIs []string `json:"redirectURIs,omitempty" description:"valid redirection URIs associated with a client"`
+
+	// GrantStrategy determines how to handle grants for this client. If no method is provided, the
+	// cluster default grant handling method will be used
+	GrantStrategy GrantHandlerType `json:"grantStrategy,omitempty"`
```
Comment (smarterclayton) [inline pkg/oauth/api/v1/types.go]: Add a longer description here, including what each strategy does (so it shows up in swagger).

**Comment by stevekuznetsov** [comment]
@liggitt time to revive this:
 - did we figure out the best way to test this? AFAIR:
  - start the OpenShift OAuth server
  - create clients with different grant strategies
  - ensure redirect happened when expected
 - do we want custom conversion funcs for v1beta3 stuff? do we need that compat?
 - do we want this PR to also persist the grants?

**Comment by deads2k** [comment]
> do we want this PR to also persist the grants?

I think the grants themselves are persisted.  This is about helping us give some client "special" powers when the cluster-admin believes they are used as part of the infrastructure.

**Comment by liggitt** [comment]
closing in favor of https://github.com/openshift/origin/pull/9616

**Pull request closed without merging**